### PR TITLE
Improve recreation of instances

### PR DESCRIPTION
**Pull request by santiago-ventura**
# Overview

This pull request introduces a series of enhancements and new features for Cloud Foundry™ Services Operator. Key improvements include the implementation of exponential back-off, enhanced error handling, and comprehensive integration tests.

## Motivation and Context

The changes are necessary to ensure higher stability and reliability of the operator under various conditions, especially in production environments. Enhanced how the CF Service Operator handles failures of Cloud Foundry operations such as service instance creation and deletion failures.

## Description of Changes

- **Exponential Back-Off**: Introduced logic for exponential back-off in service instance recreation, enhancing reconciliation reliability.
- **Retry Limits**: Established a maximum number of retries for service instance recreation to prevent endless loops by introducing an annotation.
- **Reconcile Timeout**: Added new logic to make the time interval between reconcile configurable as annotations
- **Test Suite**:
  - Added integration tests using `suite_test`.
  - Introduce the counterfeiter library to mock CF API request.
- **Integration Tests**: Developed specific integration tests to validate different scenarios for the creation and re-creation of CF Service instances and spaces.
- **Makefile Changes**:
  - `CRD Generation`, Ensured all necessary CRD files are automatically generated when `make test-fast` is executed.
  - A new make command `clean` has been introduced to remove binary libraries that do not match the actual tool library versions.
  - Updated Controller Tool library version to `v0.14.0`.
- **Linting Support**: Implemented linting using ESLint to enforce code quality standards.
- **Documentation**: Updated detailed documentation for annotations related to these new features within the operator.

## How Has This Been Tested?

The changes have been tested on Kubernetes in development and staging environments.

## Types of changes

- [x] New feature
- [x] Tests
- [x] Documentation update

## Checklist

- [x] My code follows the code style of this project.
- [x] I have added tests to cover my changes.
- [x] All new tests passed.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: internal/controllers/serviceinstance_controller_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+			// CreateInstance shall always fail directly
+			fakeSpaceClient.CreateInstanceReturns(errCreateInstanceFail)
+
+			// GetInstance shall return errors except for below cases
+			fakeSpaceClient.GetInstanceReturns(kNoInstance, errNotExpected)
+			fakeSpaceClient.GetInstanceReturnsOnCall(0, &fakeInstanceFailed, kNoError) // Instance creation fails
+			for i := 1; i <= 8; i++ {
+				fakeSpaceClient.GetInstanceReturnsOnCall(i, kNoInstance, kNoError)
+			}
```
[the remark this answers is not in the document]
Comment (shilparamasamyreddy) [reply internal/controllers/serviceinstance_controller_integration_test.go]: Hi Uwe,

I have adopted the suggestions in https://github.com/SAP/cf-service-operator/pull/48/commits/5f95d0cad05ff3d02efcfe60467268624850c802. 
Please review. Thanks.

**Comment by TheBigElmo** [comment]
I couldn't find the logic behind the **Exponential Back-Off** feature. Could you please guide me?
If exponential back-off is achieved, it will be possible to set the maximum requeue time for failing resources to 30 or 60 minutes. This would eliminate the need for a maximum retry logic and the controller would remain the model of eventual consistency of its resources.

**Comment by santiago-ventura** [comment]
> I couldn't find the logic behind the **Exponential Back-Off** feature. Could you please guide me? If exponential back-off is achieved, it will be possible to set the maximum requeue time for failing resources to something like 30 or 60 minutes. This would eliminate the need for a maximum retry logic and the controller would remain the model of eventual consistency of its resources.

Hi, we have included the logic for the **Exponential Back-Off** into the [HandleError ](https://github.com/SAP/cf-service-operator/blob/153ee101bb5771c774134069ab908997303332d4/internal/controllers/serviceinstance_controller.go#L468) function.

**Comment by santiago-ventura** [comment]
> I couldn't find the logic behind the **Exponential Back-Off** feature. Could you please guide me? If exponential back-off is achieved, it will be possible to set the maximum requeue time for failing resources to something like 30 or 60 minutes. This would eliminate the need for a maximum retry logic and the controller would remain the model of eventual consistency of its resources.

Hi colleague,

I understand your idea on setting a maximum requeue time for failing resources, but a maximum retry logic is a common practice as well. Exponential back-off helps manage the load by increasing delays between retries, which is useful in temporary failures. However, without a maximum retry limit, a failing resource might be continuously retried without resolution, especially if the error is serious and persistent.

The retry logic and retry enable the operator to signal that recovery may be impossible after numerous attempts, such as when 10 retries fail to resolve an issue. This avoids resource wastage and helps identify problems needing manual resolution.

Without a maximum retry, errors could trap resources in endless retries, preventing them from reaching a final state.

File: internal/controllers/serviceinstance_controller_integration_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (RalfHammer) [reply internal/controllers/serviceinstance_controller_integration_test.go]: The original idea here was to let every call fail that is not expected.
The only exceptions were listed below

call number 0 should return failed instance
calls 1 to 8 should return empty instance
The new version of the test code will now just return no instance without an error.
I would suggest to revert to original coding.

**Comment by TheBigElmo** [comment]
> > I couldn't find the logic behind the **Exponential Back-Off** feature. Could you please guide me? If exponential back-off is achieved, it will be possible to set the maximum requeue time for failing resources to something like 30 or 60 minutes. This would eliminate the need for a maximum retry logic and the controller would remain the model of eventual consistency of its resources.
> 
> Hi colleague,
> 
> I understand your idea on setting a maximum requeue time for failing resources, but a maximum retry logic is a common practice as well. Exponential back-off helps manage the load by increasing delays between retries, which is useful in temporary failures. However, without a maximum retry limit, a failing resource might be continuously retried without resolution, especially if the error is serious and persistent.
> 
> The retry logic and retry enable the operator to signal that recovery may be impossible after numerous attempts, such as when 10 retries fail to resolve an issue. This avoids resource wastage and helps identify problems needing manual resolution.
> 
> Without a maximum retry, errors could trap resources in endless retries, preventing them from reaching a final state.

Hi @santiago-ventura,

In Kubernetes, maximum retry logic isn't common. The problem is that resources will never be ready without manual interaction. The other question is, why should you keep a failed instance? Either you expect the resource to be ready at some point, or it won't ever work, but then the best practice is to clean up such a resource.
Which leads to the following suggestion: if you still need such a behavior, please disable it by default and make it possible to start the operator with an additional parameter to enable it and allow to override the default maximum number of retries.

**Comment by bKiralyWdf** [comment]
> > > I couldn't find the logic behind the **Exponential Back-Off** feature. Could you please guide me? If exponential back-off is achieved, it will be possible to set the maximum requeue time for failing resources to something like 30 or 60 minutes. This would eliminate the need for a maximum retry logic and the controller would remain the model of eventual consistency of its resources.
> > 
> > 
> > Hi colleague,
> > I understand your idea on setting a maximum requeue time for failing resources, but a maximum retry logic is a common practice as well. Exponential back-off helps manage the load by increasing delays between retries, which is useful in temporary failures. However, without a maximum retry limit, a failing resource might be continuously retried without resolution, especially if the error is serious and persistent.
> > The retry logic and retry enable the operator to signal that recovery may be impossible after numerous attempts, such as when 10 retries fail to resolve an issue. This avoids resource wastage and helps identify problems needing manual resolution.
> > Without a maximum retry, errors could trap resources in endless retries, preventing them from reaching a final state.
> 
> Hi @santiago-ventura,
> 
> In Kubernetes, maximum retry logic isn't common. The problem is that resources will never be ready without manual interaction. The other question is, why should you keep a failed instance? Either you expect the resource to be ready at some point, or it won't ever work, but then the best practice is to clean up such a resource. Which leads to the following suggestion: if you still need such a behavior, please disable it by default and make it possible to start the operator with an additional parameter to enable it and allow to override the default maximum number of retries.

I agree that we should keep the default behaviour, and only cap retries if the consumer actively wants this (i.e. sets the appropriate annotation on the ressource). I would not want to set this on an operator level, because it is possible that for some instances more retries are necessary (e.g. less reliable CF service). So I would say the default behaviour defined here https://github.com/SAP/cf-service-operator/pull/48/files#diff-10aec153ceb471e285d9e444db581f8efd500c63fecf207f114d1ab4b746b451R47 among others should be changed to "infinite retries". @santiago-ventura can you please make that change?

 This is analogous to how Kubernetes handles job backoff limits (https://kubernetes.io/docs/tasks/job/pod-failure-policy/, https://kubernetes.io/docs/concepts/workloads/controllers/job/#pod-failure-policy)

**Comment by santiago-ventura** [comment]
Hi @TheBigElmo, 

We followed your suggestion and kept the infinite retries behavior as default.
The Maximum Retries and only enable via annotation.

https://github.com/SAP/cf-service-operator/blob/2d113bcf551cfaaa4a54f8f367362a19e5929e99/internal/controllers/serviceinstance_controller.go#L48

Thanks

**Comment by bKiralyWdf** [comment]
> > Hi @TheBigElmo,
> > We followed your suggestion and kept the infinite retries behavior as default. The Maximum Retries and only enable via annotation.
> > https://github.com/SAP/cf-service-operator/blob/2d113bcf551cfaaa4a54f8f367362a19e5929e99/internal/controllers/serviceinstance_controller.go#L48
> > 
> > Thanks
> 
> Can we set the default to - 1? So we don't change the default behavior.

That would be overwritten by this https://github.com/SAP/cf-service-operator/pull/48/commits/14cb1322a0e5306974ae3fc1650b6e0dfecc7596#diff-10aec153ceb471e285d9e444db581f8efd500c63fecf207f114d1ab4b746b451R421

The current code keeps the default behaviour of infinite retries, because https://github.com/SAP/cf-service-operator/pull/48/commits/14cb1322a0e5306974ae3fc1650b6e0dfecc7596#diff-10aec153ceb471e285d9e444db581f8efd500c63fecf207f114d1ab4b746b451R461 checks if the retry count is changed. This functionally means that a value of `math.MaxInt32` would retry indefinitely, not just `math.MaxInt32` times.

**Review by TheBigElmo — APPROVED** [review]

**Pull request merged**